### PR TITLE
Factor out the Harness changes for test_model.py

### PR DESCRIPTION
**Pull request by jameinel**
If you want to call something like Model.relations['db'] when there
isn't a relation established, we need to have CharmMeta in the
ModelBackend so we give the right errors.

Also, track all the calls to the model backend, so when doing Model tests, we can validate that we interact with the backend in the way we expect.

### CODE REVIEW

File: ops/testing.py [hunk 1]
```diff
@@ -140,7 +140,7 @@ def _next_relation_id(self):
         self._relation_id_counter += 1
         return rel_id
 
-    def add_relation(self, relation_name, remote_app, remote_app_data=None):
+    def add_relation(self, relation_name, remote_app, *, remote_app_data=None):
```
Comment (facundobatista) [inline ops/testing.py]: Is this change tested? (same with similar change below)

(BTW, I think it's a good idea to force kwargs parameters when it's too confusing to be used as positional)

File: ops/testing.py [hunk 2]
```diff
@@ -319,16 +319,33 @@ def set_leader(self, is_leader=True):
...
+        """Return the calls that we have made to the TestingModelBackend.
+
+        This is useful mostly for testing the framework itself, so that we can assert that we
+        do/don't trigger extra calls.
+
+        :param reset: If True, reset the calls list back to empty.
+        :type reset: bool
+        :return: [(call1, args...), (call2, args...)]
+        """
+        calls = self._backend._calls.copy()
```
Comment (facundobatista) [inline ops/testing.py]: We shouldn't use private attributes from other objects.

Or the backend should allow us to use the attribute (make it public), or provide a way to access those calls.

Don't be shy on making `calls` a public attribute, it's in a private class anyway so it shouldn't be used from outside the harness anyway...

File: ops/testing.py [hunk 3]
```diff
@@ -361,52 +392,69 @@ def relation_set(self, relation_id, key, value, is_app):
         if bucket_key not in relation:
             relation[bucket_key] = {}
         bucket = relation[bucket_key]
-        bucket[key] = value
+        if value == '':
+            bucket.pop(key, None)
+        else:
+            bucket[key] = value
```
Comment (facundobatista) [inline ops/testing.py]: Is this change related to this branch?

File: ops/testing.py [hunk 4]
```diff
@@ -342,17 +359,31 @@ def __init__(self, unit_name):
         self._unit_status = None
 
     def relation_ids(self, relation_name):
-        return self._relation_ids_map[relation_name]
+        self._calls.append(('relation_ids', relation_name))
```
Comment (facundobatista) [inline ops/testing.py]: It would be awesome if we could record the call for each method without having to copy this line for eeeeeeeevery method.

What about using a class decorator? is this too much magic? http://linkode.org/#TxUe0WImzhIRbORBemRlL

File: ops/testing.py [hunk 5]
```diff
@@ -342,17 +359,31 @@ def __init__(self, unit_name):
...
+        try:
+            return self._relation_ids_map[relation_name]
+        except KeyError as e:
+            if relation_name not in self._meta.relations:
+                raise model.ModelError('{} is not a known relation'.format(relation_name)) from e
+            return []
```
Comment (dshcherb) [inline ops/testing.py:368]: Thanks for this.

File: ops/testing.py (hunk 2 above)
Comment (jameinel) [reply ops/testing.py:332]: So, the model that I used was that _TestModelBackend's public API is strictly conforming to what Model needs from ModelBackend, and than Harness was the place that kept the consistency of the _TestModelBackend's internal structures.
We *can* add a bunch of public methods on _TestModelBackend, but that confuses what exists to be a ModelBackend, and what exists to be a backend-that-can-be-mutated-by-Harness.

If we want to change that pattern, I'm open to the discussion. But I really did want _TestModelBackend to be mostly dumb, and Harness to be the glue that kept it consistent. (put another way, the consistency is in Harness, and _TestModelBackend is just an implementation detail of *Harness* that is passed to Model.)

File: ops/testing.py (hunk 3 above)
Comment (jameinel) [reply ops/testing.py:399]: There is a test added related to this change (test_relation_set deletes)
It was something that I encountered as part of working on test_model.py
This branch was a collection of changes to Harness that were necessary to do other testing, but they weren't independently factored apart.

File: ops/testing.py (hunk 1 above)
Comment (jameinel) [reply ops/testing.py:143]: I didn't add a test that "you cannot pass this as an *arg". I did find that in all the places I was using it, I was using a named arg because it felt a lot cleaner, so I wanted to recommend that to others. If you feel strongly I can have the:
```
with self.assertRaises(TypeError) as cm:
  harness.add_relation('name', 'app', {'app': 'data'})
```
it doesn't seem like the sort of test I would have added if I had written the code that way from the beginning, though.

File: ops/testing.py (hunk 2 above)
Comment (jameinel) [reply ops/testing.py:332]: I should also note that we had some discussions about accessing private methods from Harness of other classes. Namely Harness also does poke at private internals of Model (which is a much bigger no-no than private internals of the coupled class in the same module).
However, the discussion between Gustavo and I ended up with "Harness is the official, supported way for 3rd parties to poke at internals of Model, rather than exposing those knobs as 'public' on Model.". (namely _invalidate() on a couple of the cache objects could be made public.)
So while it overly ties Harness to the other Framework classes, since that is the purpose of Harness, we can move forward with it.
It is true that some aspects of Harness are going to be more brittle than they might be otherwise, but it also avoids having to publish "public for others" attributes for things that we want "private to the framework, but supported between framework objects.".

Comment (facundobatista) [reply ops/testing.py:332]: Perfect. May you please improve the _TestModelBackend docstring with some of this information? Specially stating that is responsibility of the Harness class to ensure consistency between `_relation_ids_map`, `_relation_names`, `_relation_list_map`, and `_relation_data`...

Thanks!!

File: ops/testing.py (hunk 3 above)
Comment (facundobatista) [reply ops/testing.py:399]: Ack

File: ops/testing.py (hunk 1 above)
Comment (facundobatista) [reply ops/testing.py:143]: Is this change providing any value to the code? If yes, it should be tested so it's not accidentally lost in a refactor.

This is particularly important as it's an interface exposed to the developer, so we need to handle this carefully (for example, this kind of change can be made now without handling interface evolution because we're alpha, but in a future this kind of change will be need to be managed differently).

Comment (jameinel) [reply ops/testing.py:143]: added tests for these

File: ops/testing.py (hunk 2 above)
Comment (jameinel) [reply ops/testing.py:332]: added.

File: ops/testing.py (hunk 4 above)
Comment (jameinel) [reply ops/testing.py:362]: I imagine you could do something like this with `__getattribute__` tricks as well. (though there you have to watch out for members vs methods.)

The one bit I don't particularly prefer is that it means args vs kwargs ends up in every method call, even where we wouldn't care whether it would be passed by name or passed by position. Which also leads to double nesting of tuples, etc.

So instead of being able to do:
```
        self.assertEqual([
            ('relation_ids', 'db'),
            ('relation_list', rel_id),
        ], harness.get_backend_calls(reset=True))
```

You end up with:
```
        self.assertEqual([
            ('relation_ids', ('db',), {}),
            ('relation_list', (rel_id,), {}),
        ], harness.get_backend_calls(reset=True))
```

Which has quite a bit more 'noise'. And suddenly that starts breaking if someone does:
```
- self.backend.relation_ids(name)
+ self.backend.relation_ids(relation_name=name)
```

Even though that isn't a functional change to what is being requested.

We could handle the visual noise with something like:
```
def decorator(orig_method):
    def wrapped(self, *args, **kwargs):
        full_args = (orig_method.__name__,) + args
        if kwargs:
            full_args = full_args + (kwargs,)
	self.calls.append(full_args)
	return orig_method(self, *args, **kwargs)
    return wrapped
```

Which in your test example gives:
```
$ py ./record_calls.py
<function record_calls.<locals>.decorator.<locals>.wrapped at 0x7ff2bac31d90>
<function record_calls.<locals>.decorator.<locals>.wrapped at 0x7ff2bac316a8>
12
123 456
[('foo', 12), ('bar', 123, {'b': 456})]
```

I'm happy enough with that, that I'll implement it. If you prefer something else we can discuss. (I'll live with the not-actually-functionally-different for now.)

File: ops/testing.py [hunk 6]
```diff
@@ -319,16 +319,33 @@ def set_leader(self, is_leader=True):
         if is_leader and not was_leader and self._charm is not None and self._hooks_enabled:
             self._charm.on.leader_elected.emit()
 
+    def get_backend_calls(self, reset=False):
```
Comment (jameinel) [inline ops/testing.py:322]: While here, I went ahead and changed the default to "reset=True'. Having written a few tests, it is usually quite a bit clearer to just list the calls since the last time you checked, rather than having a growing list that says [('a',)] then [('a',),('b',)], then ([('a',), ('b',), ('c'),)]

best to change it before anyone depends on it.

File: ops/testing.py [hunk 7]
```diff
@@ -173,7 +173,7 @@ def add_relation(self, relation_name, remote_app, remote_app_data=None):
         #       remote_app_data isn't empty.
         return rel_id
 
-    def add_relation_unit(self, relation_id, remote_unit_name, remote_unit_data=None):
+    def add_relation_unit(self, relation_id, remote_unit_name, *, remote_unit_data=None):
```
Comment (niemeyer) [inline ops/testing.py:176]: We need to discuss the idea of these extra parameters, per discussion on that other PR, sooner rather than later, otherwise people will start trusting the parameters and we will drive people mad by breaking their tests.

File: ops/testing.py [hunk 8]
```diff
@@ -319,16 +319,60 @@ def set_leader(self, is_leader=True):
         if is_leader and not was_leader and self._charm is not None and self._hooks_enabled:
             self._charm.on.leader_elected.emit()
 
+    def get_backend_calls(self, reset=True):
+        """Return the calls that we have made to the TestingModelBackend.
```
Comment (niemeyer) [inline ops/testing.py:323]: It seems a bit premature to stick that in our exposed API. We should discourage the idea of testing by checking what exact methods the implementation has called, and should instead encourage testing by checking that the expected behavior did take place.  This is strongly encouraging the former.

File: ops/testing.py [hunk 9]
```diff
@@ -319,16 +319,60 @@ def set_leader(self, is_leader=True):
...
+        :return: [(call1, args...), (call2, args...)]
+        """
+        calls = self._backend._calls.copy()
+        if reset:
+            self._backend._calls.clear()
+        return calls
+
+
+def record_calls(cls):
+    """Replace methods on cls with methods that record that they have been called.
```
Comment (niemeyer) [inline ops/testing.py:340]: Same idea.

File: ops/testing.py [hunk 10]
```diff
@@ -319,16 +319,60 @@ def set_leader(self, is_leader=True):
...
+        setattr(cls, meth_name, decorator(orig_method))
+    return cls
+
+
+@record_calls
 class _TestingModelBackend:
     """This conforms to the interface for ModelBackend but provides canned data.
 
-    You should not use this class directly, it is used by `TestingHarness`_ to drive the model.
+    You should not use this class directly, it is used by `Harness`_ to drive the model.
```
Comment (niemeyer) [inline ops/testing.py:366]: s/You should not/Do NOT/

File: ops/testing.py [hunk 11]
```diff
@@ -342,14 +386,24 @@ def __init__(self, unit_name):
         self._unit_status = None
 
     def relation_ids(self, relation_name):
-        return self._relation_ids_map[relation_name]
+        try:
+            return self._relation_ids_map[relation_name]
+        except KeyError as e:
+            if relation_name not in self._meta.relations:
+                raise model.ModelError('{} is not a known relation'.format(relation_name)) from e
```
Comment (niemeyer) [inline ops/testing.py:393]: Nice.

File: ops/testing.py [hunk 12]
```diff
@@ -342,14 +386,24 @@ def __init__(self, unit_name):
...
+            return []
 
     def relation_list(self, relation_id):
-        return self._relation_list_map[relation_id]
+        try:
+            return self._relation_list_map[relation_id]
+        except KeyError as e:
+            raise model.RelationNotFoundError from e
```
Comment (niemeyer) [inline ops/testing.py:400]: Again, nice to see the model being tuned to reflect the real implementation. I suspect we'll have a lot of that in the near future.

File: test/test_testing.py [hunk 13]
```diff
@@ -49,6 +51,25 @@ def test_add_relation(self):
         self.assertEqual([rel_id], backend.relation_ids('db'))
         self.assertEqual([], backend.relation_list(rel_id))
 
+    def test_add_relation_must_name_app_data(self):
+        # language=YAML
```
Comment (niemeyer) [inline test/test_testing.py:55]: Strange.. what are those about? They don't feel so great here.

File: test/test_testing.py [hunk 14]
```diff
@@ -258,6 +299,80 @@ def cleanup():
...
+            name: test-charm
+            requires:
+                db:
+                    interface: pgsql
+            ''')
+        harness.begin()
+        harness.set_leader(False)
+        rel_id = harness.add_relation('db', 'postgresql')
+        harness.update_relation_data(rel_id, 'test-charm/0', {'foo': 'bar'})
+        harness.add_relation_unit(rel_id, 'postgresql/0')
```
Comment (niemeyer) [inline test/test_testing.py:314]: Those interactions via the harness are great! So readable and so clear.

File: test/test_testing.py [hunk 15]
```diff
@@ -258,6 +299,80 @@ def cleanup():
...
+        harness.begin()
+        # No calls to the backend yet
+        self.assertEqual([], harness.get_backend_calls())
+        rel_id = harness.add_relation('db', 'postgresql')
+        # update_relation_data ensures the cached data for the relation is wiped
+        harness.update_relation_data(rel_id, 'test-charm/0', {'foo': 'bar'})
+        self.assertEqual([
+            ('relation_ids', 'db'),
+            ('relation_list', rel_id),
+        ], harness.get_backend_calls(reset=True))
```
Comment (niemeyer) [inline test/test_testing.py:336]: This is the awkward bit.. what just happened there?  How and why do we constrain the above logic to that list?  Why does it matter?

**Comment by gnuoy** [comment]
I have tested this PR as a potential fix for issue #241 and it does indeed fix it.

**Comment by jameinel** [comment]
Closing this for now, as it isn't making use of the _get_backend_calls (which are intended for test_model.py). I'll bring it back once that makes it clear how this helps.
